### PR TITLE
Added reference tests to ReductionOpInfo and ported mean test

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63978
* __->__ #62900

fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

### CODE REVIEW

File: test/test_reductions.py [hunk 1]
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
...
+            t = make_tensor(shape, device, dtype, noncontiguous=True)
+            for kwargs in _generate_reduction_kwargs(ndim, op.supports_multiple_dims):
+                dim = kwargs.pop('dim')
+                keepdim = kwargs.pop('keepdim')
+                for args, kwargs in op.generate_args_kwargs(t, dim=dim):
+                    res = op(t, *args, dim=dim, keepdim=keepdim, **kwargs)
+                    ref = op(t.contiguous(), *args, dim=dim, keepdim=keepdim, **kwargs)
+                    self.assertEqual(res, ref)
+
+    # NumPy does not support BFloat16 so we don't test that agains
```
Comment (mruberry) [inline test/test_reductions.py:260]: Nice comment

File: test/test_reductions.py [hunk 2]
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
...
+                for args, kwargs in op.generate_args_kwargs(t, dim=dim):
+                    res = op(t, *args, dim=dim, keepdim=keepdim, **kwargs)
+                    ref = op.reference(t, *args, dim=dim, keepdim=keepdim, **kwargs)
+                    self.assertEqual(res, ref, exact_dtype=False)
+
+    @ops(filter(lambda op: op.reference is not None, reduction_op_db),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_random_input_large(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing large tensors for accuracy"""
+        t = make_tensor([10000000], device, dtype)
```
Comment (mruberry) [inline test/test_reductions.py:294]: power of 2, maybe?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -5212,6 +5218,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
 
 
+def reference_reduction_numpy(f):
+    @wraps(f)
+    def wrapper(t: torch.Tensor, *args, dim=None, keepdim=False, **kwargs):
+        # NumPy reductions don't accept dim=0 for scalar inputs
+        if t.ndim == 0 and (dim == 0 or dim == -1):
+            dim = None
+
+        return f(t.cpu().numpy(), *args, axis=dim, keepdims=keepdim, **kwargs)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5228]: .detach().cpu().numpy()?

**Review by mruberry — COMMENTED** [review]
Just a few comments/suggestions inline. The thorough expansion of testing is great. Test time is always a looming concern, however. When all the new tests are added and old tests made redundant are removed we should diff the before/after e2e ci time.

File: test/test_reductions.py [hunk 4]
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
         result = op(t, *args, **kwargs)
         self.assertEqual(result, op(t[~t.isnan()], *args, **kwargs))
 
+    @ops(reduction_op_db)
+    def test_non_contiguous_input(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing along non contiguous dimensions"""
+        for ndim in range(1, 4):
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply test/test_reductions.py:249]: Revised to manually generate some test cases

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -906,6 +910,7 @@ def sample_inputs_func(*args, **kwargs):
             return result
 
         # Override OpInfo defaults and call base class __init__
+        kwargs['ref'] = None
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:913]: Updated to use OpInfo ref

File: test/test_reductions.py [hunk 6]
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
...
+                    ref = op(t.contiguous(), *args, dim=dim, keepdim=keepdim, **kwargs)
+                    self.assertEqual(res, ref)
+
+    # NumPy does not support BFloat16 so we don't test that agains
+    # reference implementations. We also don't compare dtypes because NumPy
+    # seems to be returning a different dtype for sum/prod on Windows. Also, we
+    # already specific test for checking output dtype.
+
+    @ops(filter(lambda op: op.reference is not None, reduction_op_db),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply test/test_reductions.py:266]: The tests are running pretty fast right now, also this is testing for the default test dtypes, if we want to constrain it we should do it through the opinfo test_dtypes kwarg.

File: test/test_reductions.py [hunk 7]
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
...
+        for args, kwargs in op.generate_args_kwargs(t_contig, **reduction_kwargs):
+            kwargs.update(reduction_kwargs)
+            result = op(t, *args, **kwargs)
+            expected = op(t_contig, *args, **kwargs)
+            self.assertEqual(result, expected)
+
+    @ops(reduction_ops)
+    def test_noncontiguous_innermost(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing along noncontiguous innermost dimension."""
+        t = make_tensor((10, 10), device, dtype)
```
Comment (mruberry) [inline test/test_reductions.py:311]: make_tensor has a noncontiguous kwarg

https://github.com/pytorch/pytorch/blob/fcc1f87b6aacdfba9193691568bc405be42ac77e/torch/testing/_internal/common_utils.py#L1942

Comment (mruberry) [reply test/test_reductions.py:311]: I guess the issue with using that is you want to control the noncontiguity precisely here

File: test/test_reductions.py [hunk 8]
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
...
+    @ops(reduction_ops)
+    def test_noncontiguous_outermost(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing along noncontiguous outermost dimension."""
+        t = make_tensor((10, 10), device, dtype)
+        self._test_noncontiguous(op, t[::2, :], dim=0)
+
+    @ops(reduction_ops)
+    def test_noncontiguous_all(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing all dimensions of a noncontiguous tensor."""
+        t = make_tensor((10, 10), device, dtype)
```
Comment (mruberry) [inline test/test_reductions.py:323]: 3D might be more interesting for this test

File: test/test_reductions.py [hunk 9]
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
...
+    def test_noncontiguous_expanded(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing a tensor with expanded singleton dimensions."""
+        t = make_tensor((10, 10), device, dtype)
+        self._test_noncontiguous(op, t.unsqueeze(1).expand(-1, 5, -1))
+
+    # NumPy does not support BFloat16 so we don't test that against reference
+    # implementations. We also don't compare dtypes or test for different
+    # keepdim because we already have other tests covering those.
+
+    def _test_ref(self, op: ReductionOpInfo, t: torch.Tensor, **reduction_kwargs):
```
Comment (mruberry) [inline test/test_reductions.py:342]: Maybe add a comment about how these reference tests are distinct from

https://github.com/pytorch/pytorch/blob/feba6806c9b7a5e13ffd5839f8bc6ddc2e016a26/test/test_ops.py#L199

which only uses sample inputs

File: test/test_reductions.py [hunk 10]
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
...
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_large_input(self, device, dtype, op: ReductionOpInfo):
+        """Compares op against reference for a very large input to check stability"""
+        self._test_ref(op, make_tensor((2 ** 20,), device, dtype))
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_extremal_values(self, device, dtype, op: ReductionOpInfo):
```
Comment (mruberry) [inline test/test_reductions.py:380]: Running this on integer types doesn't seem interesting. In the interest of test time float32 and complex64 is probably OK?

File: test/test_reductions.py [hunk 11]
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
...
+        self._test_ref(op, make(32))
+        self._test_ref(op, make(5, 2, 4, 3))
+        self._test_ref(op, make(5, 2, 4, 3), dim=0)
+        self._test_ref(op, make(5, 2, 4, 3), dim=-1)
+
+        if op.supports_multiple_dims:
+            self._test_ref(op, make(10, 1, 7, 4), dim=(0, -1))
+            self._test_ref(op, make(10, 1, 7, 4), dim=(0, 1, 3))
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
```
Comment (mruberry) [inline test/test_reductions.py:372]: Maybe a subset of dtypes, like uint8, int16, long, float16, float32, and complex64 would alleviate test time?

File: test/test_reductions.py [hunk 12]
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
...
+
+    def _test_ref(self, op: ReductionOpInfo, t: torch.Tensor, **reduction_kwargs):
+        """Compares op against op.ref for the given input and reduction kwargs"""
+        for args, kwargs in op.generate_args_kwargs(t, **reduction_kwargs):
+            kwargs.update(reduction_kwargs)
+            result = op(t, *args, **kwargs)
+            expected = op.ref(t.detach().cpu().numpy(), *args, **kwargs)
+            self.assertEqual(result, expected, exact_dtype=False)
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
```
Comment (mruberry) [inline test/test_reductions.py:350]: See note about running on a subset of dtypes below

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -5384,6 +5386,53 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     return op(input.triu() if upper else input.tril(), upper)
 
 
+def reference_reduction_numpy(f, supports_keepdims=True):
+    """Wraps a NumPy reduction operator.
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5390]: Nice docstring

**Review by mruberry — APPROVED** [review]
Just a few small suggestions inline for your review

File: test/test_reductions.py (hunk 7 above)
Comment (heitorschueroff) [reply test/test_reductions.py:311]: Yeah, in this case I'd like to test for different types of non-contiguity

File: test/test_reductions.py (hunk 11 above)
Comment (heitorschueroff) [reply test/test_reductions.py:372]: I set the default_test_dtypes to just a subset and by default @ops uses the default_test_dtypes if it is set.

Comment (mruberry) [reply test/test_reductions.py:372]: Yes but I mean for these reference tests in particular it might be useful to winnow the dtypes down (and not to the default set because it contains bfloat16)

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).
